src/lib.rs

```diff
@@ -1021,6 +1045,337 @@ safety_comment! {
+
+    // SAFETY: `[T]`'s trailing slice is `[T]`, which is trivially aligned.
+    unsafe impl<T> SplitAt for [T] {
+        type Elem = T;
+
+        #[inline]
+        fn only_derive_is_allowed_to_implement_this_trait()
+        where
+            Self: Sized,
+        {
+        }
+    }
+
+    /// Splits `T` in two.
+    ///
+    /// # Safety
+    ///
+    /// The caller promises that:
+    /// 0. `l_len.get()` is not greater than the length of `ptr`'s trailing
+    ///    slice.
+    /// 1. if `I::Aliasing` is [`Exclusive`] or `T` permits interior mutation,
+    ///    then `l_len.padding_needed_for() == 0`.
+    #[inline(always)]
+    unsafe fn ptr_split_at_unchecked<'a, T, I, R>(
+        ptr: Ptr<'a, T, I>,
+        l_len: MetadataOf<T>,
+    ) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+    where
+        I: invariant::Invariants,
+        T: ?Sized + pointer::Read<I::Aliasing, R> + SplitAt,
+    {
+        let inner = ptr.as_inner();
+
+        // SAFETY: The caller promises that `l_len.get()` is not greater than
+        // the length of `self`'s trailing slice.
+        let (left, right) = unsafe { inner.split_at_unchecked(l_len) };
+
+        // Lemma 0: `left` and `right` conform to the aliasing invariant
+        // `I::Aliasing`. Proof: If `I::Aliasing` is `Exclusive`, the caller
+        // promises that `l_len.padding_needed_for() == 0`. Consequently,
+        // there is no trailing padding after `left`'s final element that would
+        // overlap into `right`. If `I::Aliasing` is shared, then overlap
+        // between their referents is permissible.
+
+        // SAFETY:
+        // 0. `left` conforms to the aliasing invariant of `I::Aliasing, by
+        //    Lemma 0.
+        // 1. `left` conforms to the alignment invariant of `I::Alignment,
+        //    because the referents of `left` and `Self` have the same address
+        //    and type (and, thus, alignment requirement).
+        // 2. `left` conforms to the validity invariant of `I::Validity``,
+        //    because TODO.
+        let left = unsafe { Ptr::from_inner(left) };
+
+        // SAFETY:
+        // 0. `right` conforms to the aliasing invariant of `I::Aliasing, by
+        //    Lemma 0.
+        // 1. `right` conforms to the alignment invariant of `I::Alignment,
+        //    because TODO. (this is a tricky one)
+        // 2. `right` conforms to the validity invariant of `I::Validity``,
+        //    because TODO. (this is a tricky one).
+        let right = unsafe { Ptr::from_inner(right) };
+
+        (left, right)
+    }
+
+    #[cfg(test)]
+    mod tests {
+        #[cfg(feature = "derive")]
+        #[test]
+        fn test_split_at() {
+            use crate::{FromBytes, Immutable, IntoBytes, KnownLayout, SplitAt};
+
+            #[derive(FromBytes, KnownLayout, SplitAt, IntoBytes, Immutable)]
+            #[repr(C)]
+            struct SliceDST<const OFFSET: usize> {
+                prefix: [u8; OFFSET],
+                trailing: [u8],
```

Comment:
We should probably also test for other trailing slice element sizes. Maybe make this generic over `T` and have `trailing: [T]` and have `test_split_at` be generic over `T`?

Reply:
This proved to be rather tricky without code duplication.

Reply:
Ack